common/argparse.h

```diff
@@ -0,0 +1,870 @@
+    // default for string arguments unless this is called.
+    //
+    // For arguments that are `Set` below, this value will be stored even if the
+    // argument is never explicitly provided. For arguments that are `Append`ed
+    // below, this value will be used whenever the argument occurs without an
+    // explicit value, but unless the argument is parsed nothing will be
+    // appended.
+    void Default(llvm::StringRef string_value);
+
+    // Configures the argument to store a parsed value in the provided storage.
+    // Each time the argument is parsed, it will write a new value to this
+    // storage.
+    //
+    // Exactly one of this method or `Append` below must be configured for the
+    // argument.
+    void Set(llvm::StringRef* string);
+
+    // Configures the argument to append a parsed value to the provided
+    // container. Each time the argument is parsed, a new value will be
+    // appended.
+    //
+    // Exactly one of this method or `Set` above must be configured for the
+    // argument.
+    void Append(llvm::SmallVectorImpl<llvm::StringRef>* sequence);
+
+   private:
+    friend Args;
+    using ArgBuilder::ArgBuilder;
+  };
+
+  // Customized argument builder when the value is required to be one of a fixed
+  // set of possible strings, and each one maps to a specific value of some
+  // type, often an enumerator.
+  class OneOfArgBuilder : public ArgBuilder {
+   public:
+    // A tiny helper / builder type for describing one of the possible values
+    // that a particular argument accepts.
+    //
+    // Beyond carrying the string, type, and value for this candidate, it also
+    // allows marking the candidate as the default.
+    template <typename T>
+    class OneOfValueT {
+     public:
+      // Configure whether a candidate is the default. If not called, it is not
+      // the default. This can only be used when setting, not when appending.
+      auto Default(bool is_default) && -> OneOfValueT;
+
+     private:
+      friend OneOfArgBuilder;
+
+      explicit OneOfValueT(llvm::StringRef str, T value);
+
+      llvm::StringRef str;
+      T value;
+      bool is_default = false;
+    };
+
+    // A helper function to create an object that models one of the candidate
+    // values for this argument. It takes the string used to select this value
+    // on the command line, deduces the type of the value, and accepts the value
+    // itself that should be used when this candidate is parsed.
+    template <typename T>
+    static auto OneOfValue(llvm::StringRef string, T value) -> OneOfValueT<T>;
+
+    // Configures the argument to store a parsed value in the provided storage.
+    // Each time the argument is parsed, it will write a new value to this
+    // storage.
+    //
+    // Exactly one of this method or `AppendOneOf` below must be configured for
+    // the argument.
+    //
+    // For one-of arguments, this method also provides an array of possible
```

Comment:
This list of requirements is a lot to parse, but looking at the examples in the test code it makes a lot more sense. Perhaps providing a short example here and for `AppendOneOf` would help? Otherwise I feel like I'm reading the disclaimer text at the end of a pharmaceutical advertisement :)

Reply:
Done. Let me know if other tweaks would help here.